exe/kubernetes-deploy

```diff
@@ -40,23 +44,29 @@ ARGV.options do |opts|
   opts.parse!
 end
 
+if use_stdin && template_dir
+  puts "The flags --use-stdin and --template-dir cannot be combined"
+  exit(1)
+end
+
 namespace = ARGV[0]
 context = ARGV[1]
-template_dir = KubernetesDeploy::OptionsHelper.default_and_check_template_dir(template_dir)
 revision = KubernetesDeploy::OptionsHelper.revision_from_environment
 logger = KubernetesDeploy::FormattedLogger.build(namespace, context, verbose_prefix: verbose_log_prefix)
 
-runner = KubernetesDeploy::DeployTask.new(
-  namespace: namespace,
-  context: context,
-  current_sha: revision,
-  template_dir: template_dir,
-  bindings: bindings,
-  logger: logger,
-  max_watch_seconds: max_watch_seconds
-)
-
 begin
+  template_dir = KubernetesDeploy::OptionsHelper.default_and_check_template_dir(template_dir, use_stdin: use_stdin)
```

Comment:
Of the two options, I think this one is a lot easier to follow overall. However, I agree that it is a better design to make the place that creates the temp dir responsible for cleaning it up--callers should not have to know what the options helper did! Maybe we could get both advantages by modifying this version to yield the validated path instead of returning it:

```ruby
begin
  KubernetesDeploy::OptionsHelper.with_validated_template_path(template_dir, use_stdin: use_stdin) do |valid_path|
    runner = KubernetesDeploy::DeployTask.new(template_dir: valid_path, ...)
    runner.run!
  end
rescue KubernetesDeploy::DeploymentTimeoutError
  exit(70)
rescue KubernetesDeploy::FatalDeploymentError
  exit(1)
end
```

```ruby
    def self.with_validated_template_dir(template_dir, use_stdin: false)
      if use_stdin
        # ...validate and make temp dir
        yield temp_dir
        return
      end

      if !template_dir && ENV.key?("ENVIRONMENT")
        yield "config/deploy/#{ENV['ENVIRONMENT']}"
      elsif !template_dir || template_dir.empty?
        raise FatalDeploymentError, "Template directory is unknown. " \
          "Either specify --template-dir argument or set $ENVIRONMENT to use config/deploy/$ENVIRONMENT " \
        + "as a default path."
      else
        yield template_dir
      end

      nil
    ensure
      FileUtils.remove_entry(temp_dir) if temp_dir
    end
```

Reply:
This turned out simpler than I thought it would be and should avoid any manual cleanup outside of a bit of error handling since we can encompass everything in the `mktmpdir` block, it will clean everything up for us on exit.